crates/flux-desugar/src/desugar.rs

```diff
@@ -470,6 +470,53 @@ impl<'a, 'genv, 'tcx: 'genv> RustItemCtxt<'a, 'genv, 'tcx> {
         Ok(fhir::ImplItem { generics, kind: fhir::ImplItemKind::Fn(fn_sig), owner_id: self.owner })
     }
 
+    pub(crate) fn desugar_const_spec(
+        &mut self,
+        def_id: LocalDefId,
+        const_info: &surface::ConstantInfo,
+    ) -> Result<fhir::ConstantInfo<'genv>> {
+        let expr = match &const_info.expr {
+            Some(expr) => Some(self.desugar_expr(expr)?),
+            None => None,
+        };
+        Ok(fhir::ConstantInfo { owner: OwnerId { def_id }, expr })
+    }
+
+    pub(crate) fn desugar_const(
+        &mut self,
+        def_id: LocalDefId,
+        const_info: &surface::ConstantInfo,
+    ) -> Result<fhir::Item<'genv>> {
+        let constant_info = self.desugar_const_spec(def_id, const_info)?;
+        let owner_id = self.owner;
+        let generics = fhir::Generics::trivial();
+        let kind = fhir::ItemKind::Const(constant_info);
+        Ok(fhir::Item { owner_id, generics, kind })
+    }
+
+    pub(crate) fn desugar_impl_const(
+        &mut self,
+        def_id: LocalDefId,
+        const_info: &surface::ConstantInfo,
+    ) -> Result<fhir::ImplItem<'genv>> {
+        let constant_info = self.desugar_const_spec(def_id, const_info)?;
+        let owner_id = self.owner;
+        let generics = fhir::Generics::trivial();
+        let kind = fhir::ImplItemKind::Const(constant_info);
+        Ok(fhir::ImplItem { owner_id, generics, kind })
+    }
+
+    pub(crate) fn desugar_trait_const(
+        &mut self,
+        def_id: LocalDefId,
+        const_info: &surface::ConstantInfo,
+    ) -> Result<fhir::TraitItem<'genv>> {
+        let constant_info = self.desugar_const_spec(def_id, const_info)?;
+        let owner_id = self.owner;
+        let generics = fhir::Generics::trivial();
+        let kind = fhir::TraitItemKind::Const(constant_info);
+        Ok(fhir::TraitItem { owner_id, generics, kind })
+    }
```

Comment:
Const in trait seems straightforward, what is the issue? I could add tests, but I think it should "just work" ....

Impl seems hairier because of the usual resolution issues -- you probably want an "associated constant" but whats the point even of that if you can't at compile time resolve the impl --> means you can't actually know what the `const` was either?

Reply:
For `impl` actually it makes sense to have them _if_ you only use them locally (i.e. not a `pub`) or I guess somehow at each usage site `rustc` can resolve it to one exact impl (otherwise, presumably it complains...?)

Reply:
NVM made `None` for now!

Reply:
Constants in traits are abstract. Implementations define them. For it to work it needs to go through normalization as we do for associated types.

Constants in inherent implementations are like item constants with the difference that they can mention generics in the impl block.